joern-cli/frontends/php2cpg/src/main/scala/io/joern/php2cpg/passes/PhpTypeRecovery.scala

```diff
@@ -0,0 +1,289 @@
+        val cpgTypes = cpg.typeDecl
+          .fullNameExact(ts.map(_.compUnitFullName).toSeq: _*)
+          .member
+          .nameExact(sym.identifier)
+          .flatMap(m => m.typeFullName +: m.dynamicTypeHintFullName)
+          .filterNot { x => x == "ANY" || x == "this" }
+          .toSet
+        if (cpgTypes.nonEmpty) cpgTypes
+        else symbolTable.get(sym)
+      case ::(head: Call, Nil) if symbolTable.contains(head) =>
+        val callPaths    = symbolTable.get(head)
+        val returnValues = methodReturnValues(callPaths.toSeq)
+        if (returnValues.isEmpty)
+          callPaths.map(c => s"$c$pathSep${XTypeRecovery.DummyReturnType}")
+        else
+          returnValues
+      case ::(head: Call, Nil) if head.argumentOut.headOption.exists(symbolTable.contains) =>
+        symbolTable
+          .get(head.argumentOut.head)
+          .map(t => Seq(t, head.name, XTypeRecovery.DummyReturnType).mkString(pathSep.toString))
+      case ::(identifier: Identifier, Nil) if symbolTable.contains(identifier) =>
+        symbolTable.get(identifier)
+      case ::(head: Call, Nil) =>
+        extractTypes(head.argument.l)
+      case _ => Set.empty
+    }
+    val returnTypes = extractTypes(ret.argumentOut.l)
+    existingTypes.addAll(returnTypes)
+
+    /* Check whether method return is already known, and if so, remove dummy value */
+    val saveTypes = existingTypes.filterNot(typeName => {
+      if (typeName.startsWith(Defines.UnresolvedNamespace))
+        true
+      else if (typeName.endsWith(s"${XTypeRecovery.DummyReturnType}"))
+        typeName.split(pathSep).headOption match {
+          case Some(methodName) => {
+            val methodReturns = methodReturnValues(Seq(methodName))
+              .filterNot(_.endsWith(s"${XTypeRecovery.DummyReturnType}"))
+            !methodReturns.isEmpty
+          }
+          case None => false
+        }
+      else
+        false
+    })
+    methodTypesTable.update(m, saveTypes)
+    builder.setNodeProperty(ret.method.methodReturn, PropertyNames.DYNAMIC_TYPE_HINT_FULL_NAME, saveTypes)
+  }
+
+  /* Necessary to change the filter regex from (this|self) to (\\$this|this), in order to account for $this PHP
+   * convention.
+   */
+  override protected def associateTypes(symbol: LocalVar, fa: FieldAccess, types: Set[String]): Set[String] = {
+    fa.astChildren.filterNot(_.code.matches("(\\$this|this|self)")).headOption.collect {
+      case fi: FieldIdentifier =>
+        getFieldParents(fa).foreach(t => persistMemberWithTypeDecl(t, fi.canonicalName, types))
+      case i: Identifier if isField(i) =>
+        getFieldParents(fa).foreach(t => persistMemberWithTypeDecl(t, i.name, types))
+    }
+    symbolTable.append(symbol, types)
+  }
+
+  /* Reference the PythonTypeRecovery implementation. The XTypeRecovery one seems incorrect. */
+  override protected def getFieldParents(fa: FieldAccess): Set[String] = {
+    if (fa.method.name == "<module>") {
```

Comment:
PHP seems to use the terminology `<global>` for module defining methods
```suggestion
    if (fa.method.name == "<global>") {
```